release-1.9: Backports for 1.9.3

KristofferC opened the pull request:
Backported PRs:
- [x] #47782 <!-- Generalize Bool parse method to AbstractString -->
- [x] #48634 <!-- Remove unused "deps" mechanism in internal sorting keywords [NFC] -->
- [x] #49931 <!-- Lock finalizers' lists at exit -->
- [x] #50064 <!-- Fix numbered prompt with input only with comment -->
- [x] #50474 <!-- docs: Fix a `!!! note` which was miscapitalized -->
- [x] #50516 <!-- Fix visibility of assert on GCC12/13 -->
- [x] #50635 <!-- `versioninfo()`: include build info and unofficial warning -->
- [x] #49915 <!-- Revert "Remove number / vector (#44358)" -->
- [x] #50781 <!-- fix `bit_map!` with aliasing -->
- [x] #50845 <!-- fix #50438, use default pool for at-threads -->
- [x] #49031 <!-- Update inference.md -->
- [x] #50289 <!-- Initialize prev_nold and nold in gc_reset_page -->
- [x] #50559 <!-- Expand kwcall lowering positional default check to vararg -->
- [x] #49582 <!-- Update HISTORY.md for `DelimitedFiles` -->
- [x] #50341 <!-- invokelatest docs should say not exported before 1.9 -->
- [x] #50525 <!-- only check that values are finite in `generic_lufact`  when `check=true` -->
- [x] #50444 <!-- Optimize getfield lowering to avoid boxing in some cases -->
- [x] #50523 <!-- Avoid generic call in most cases for getproperty -->
- [x] #50860 <!-- Add `Base.get_extension` to docs/API -->
- [x] #50164 <!-- codegen: handle dead code with unsafe_store of FCA pointers -->
- [x] #50568 <!-- `Array(::AbstractRange)` should return an `Array` -->
- [x] #50871 <!-- macOS: Don't inspect dead threadtls during exception handling. -->

Need manual backport:
- [ ] #48542 <!-- Add docs on task-specific buffering using multithreading -->
- [ ] #50591 <!-- build: fix various makefile bugs -->


Non-merged PRs with backport label:
- [ ] #50842 <!-- Avoid race conditions with recursive rm -->
- [ ] #50823 <!-- Make ranges more robust with unsigned indexes. -->
- [ ] #50663 <!-- Fix Expr(:loopinfo) codegen -->
- [ ] #49716 <!-- Update varinfo() docstring signature -->
- [ ] #49713 <!-- prevent REPL from erroring in numbered mode in some situations -->
- [ ] #49573 <!-- Implement jl_cpu_pause on PPC64 -->
- [ ] #48726 <!-- fix macro expansion of property destructuring -->
- [ ] #48642 <!-- Use gc alloc instead of alloc typed in lowering -->
- [ ] #48183 <!-- Don't use pkgimage for package if any includes fall in tracked path for coverage or alloc tracking -->
- [ ] #48050 <!-- improve `--heap-size-hint` arg handling -->
- [ ] #47615 <!-- Allow threadsafe access to buffer of type inference profiling trees -->

quinnj committed:
Generalize Bool parse method to AbstractString (#47782)

* Generalize Bool parse method to AbstractString

Fixes https://github.com/JuliaStrings/InlineStrings.jl/issues/57.

We currently have a specialization for `parse(Bool, ::Union{String, SubString{String})`
where `true` and `false` are parsed appropriately. The restriction to
`Union{String, SubString{String}}`, however, means we don't get this behavior
for other `AbstractString`s. In the linked issue above, for InlineStrings, we end up
going through the generic integer parsing codepath which results in an `InexactError`
when we try to do `Bool(10)`.

The proposal in this PR takes advantage of the fact that there is only the 2 comparisons
where we do `_memcmp` that require the input string to be "dense" (in memory), and otherwise,
we just do a comparison against a `SubString` of the input string.

Relatedly, I've wanted to introduce the concept of an abstrac type like:

```julia
abstract type MemoryAddressableString <: AbstractString
```

where the additional required interface would be being able to call
`pointer(::MemoryAddressableString)`, since a lot of our string algorithms depend on doing
these kind of pointer operations and hence makes it quite a pain to implement your own
custom string type.

* Apply suggestions from code review

Co-authored-by: Stefan Karpinski <stefan@karpinski.org>
Co-authored-by: Nick Robinson <npr251@gmail.com>

Co-authored-by: Stefan Karpinski <stefan@karpinski.org>
Co-authored-by: Nick Robinson <npr251@gmail.com>
(cherry picked from commit 63830a6f2050e61b7b2aca78e2462487fd3f59d0)

matthias314 committed:
Faster iteration over non-offset `AbstractVector` (#48720)

This change simplifies the boundscheck in loop as LLVM would lift the const subtraction.
Simd block would be generated in more cases.

Co-authored-by: N5N3 <2642243996@qq.com>
(cherry picked from commit 1543cdd5ffc734cae640e45b9bb6ada73dd81c02)

ferrolho committed:
Update inference.md (#49031)

Update URLs to JuliaHub blog

(cherry picked from commit abae35c00f55446a1d31c72090829e61a744fd17)

kpamnany committed:
Lock finalizers lists at exit (#49931)

(cherry picked from commit c470dc369865cf2f90ad34e25e110b72dd7fbd87)

vtjnash committed:
codegen: handle dead code with unsafe_store of FCA pointers (#50164)

Fix #50125

(cherry picked from commit ef6d900d441ae37313ec6711f05ea5953a0c4ab8)

d-netto committed:
initialize prev_nold and nold in gc_reset_page (#50289)

(cherry picked from commit 48c92c13573c764bde9b877e52435b757ee479ee)

stevengj committed:
invokelatest docs should say not exported before 1.9 (#50341)

(cherry picked from commit cb6d0f2e53004c4a98988dcb8b89048026eecff9)

wangl-cc committed:
Check input expresion in numbered prompt (#50064)

(cherry picked from commit 7e3c706a879ae6d381737e13ce9b3275d083a86a)

chipbuster committed:
docs: Fix a `!!! note` which was miscapitalized (#50474)

(cherry picked from commit 236c23b7945a578dd630cdf6af81ecbf7aafacf6)

gbaraldi committed:
Backport: Optimize getfield lowering to avoid boxing

vchuravy committed:
Fix visibility of assert on GCC12/13

(cherry picked from commit 28720ff2509cf4566c5f8e103be2615a3d37a380)

oscardssmith committed:
Merge pull request #50525 from oscardssmith/fix-generic_lu_fact-check=false

only check that values are finite in `generic_lufact`  when `check=true`

(cherry picked from commit 824cdf1231051189d9ee917fd48139670590c13f)

Keno committed:
Expand kwcall lowering positional default check to vararg

Fixes the case from #50518, but we actually have two test cases in
the tests that also hit this (e.g. this one:
```
f40964(xs::Int...=1; k = 2) = (xs, k)
```), but just happened not to hit the bad codegen path. #50556,
once merged would have complained on those definitions as well,
without this fix.

(cherry picked from commit c27223607daaeee94ae477a1fdeab9b8784c9a8d)

jishnub committed:
`Array(::AbstractRange)` should return an `Array` (#50568)

Currently, `Array(r::AbstractRange)` falls back to `vcat(r)`, but
certain ranges may choose to specialize `vcat(r::AbstractRange)` to not
return an `Array`. This PR ensures that `Array(r)` always returns an
`Array`.

At present, there's some code overlap with `vcat` (just above the
`Array` method added in this PR). Perhaps some of these may be replaced
by `unsafe_copyto!`, but the tests for ranges include some special cases
that don't support `getindex`, which complicates things a bit. I've not
done this for now. In any case, the common bit of code is pretty simple,
so perhaps the duplication is harmless.

(cherry picked from commit 3cc05908a61aec885f3a4fc3fc9cb13cf3524108)

IanButterworth committed:
`versioninfo()`: include build info and unofficial warning (#50635)

(cherry picked from commit c43e5a10be27b7f93b5368875aa1d2596b4d4947)

kpamnany committed:
Backport: Avoid generic call in most cases for getproperty (#50523)

(cherry picked from commit 462ac492e19df146a493536d21009f0709935551)

oxinabox committed:
Revert "Remove number / vector (#44358)" (#49915)

(cherry picked from commit 503d5b427bc246ad2a21814b94d418fd7b22e993)

oscardssmith committed:
fix `bit_map!` with aliasing (#50781)

fixes #50780 caused by #47013.

(cherry picked from commit 3e04129d61e19fe2957680b39e03b350db8e8c0d)

JeffBezanson committed:
fix #50438, use default pool for at-threads (#50845)

(cherry picked from commit 744aa79c2c2655de526f02dcd1987ae1d7db6f6d)

SBuercklin committed:
Add `Base.get_extension` to docs/API (#50860)

The documentation of `Pkg.jl` is instructing
developers to rely on the existence of this function, and there doesn't
seem to be any alternative which is a part of the API that developers
can use instead on to guarantee forward compatibility for the same
behavior.

`Base.get_extension` is [referred to
explicitly](https://pkgdocs.julialang.org/v1.9/creating-packages/#Backwards-compatibility)
in the `Pkg.jl` docs to conditionally use package extensions vs
`Requires.jl`.

The `Pkg.jl` docs suggest
```julia
if !isdefined(Base, :get_extension)
  include("../ext/PlottingContourExt.jl")
end
```
to transition from "normal dependency to extension," which will break
and automatically load the extension in future versions should
`Base.get_extension` go away.

`Base.get_extension` is the only way (that I know of) to directly access
the module associated with a package extension, which can be a useful
utility as well.

(cherry picked from commit d1759bcde8e662e6bb7aaff94466b3ca1f3cf43b)

KristofferC committed:
Update NEWS.md for DelimitedFiles

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.9")`

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":release-1.9")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50507#issuecomment-1672989477) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/606951c_vs_6fc1be0/report.md).

maleadt committed:
macOS: Don't inspect dead threadtls during exception handling. (#50871)

Otherwise we may end up with the state of a dead thread, as the system
IDs can alias. This can lead to an early return from the exception
handler, resulting in e.g. safepoint exceptions being actually delivered
to user code.

---------

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
Co-authored-by: Cody Tapscott <topolarity@tapscott.me>

imciner2 commented:
Is it possible to also backport https://github.com/JuliaLang/julia/pull/48909 to 1.9? The docs for 1.9.2 also show the issue with the admonition not appearing properly.

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50507#issuecomment-1675099132) has completed - no new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ab27a43_vs_6fc1be0/report.html) is available.
